### PR TITLE
Recover from common if let syntax mistakes/typos

### LINKED ISSUE #103587: Recover from common `if let` syntax mistakes/typos
Given the following code: [link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=56d8d05130b8a46c94a014811061b0a4)

```rust
fn main() {
    let x = Some(123);

    // #1
    if let Some(_) == x {}

    // #2
    if let x = Some(_) {}

    // #3
    if let x == Some(_) {}

    // #4
    if Some(_) = x {}
}
```

The current output is:

```
   Compiling playground v0.0.1 (/playground)
error: expected one of `=` or `|`, found `==`
 --> src/main.rs:5:20
  |
5 |     if let Some(_) == x {}
  |                    ^^ expected one of `=` or `|`

error: could not compile `playground` due to previous error
```

Ideally the output should suggest how to fix the syntax mistakes/typos for each of the above examples.
1 -> Remove extra `=`
2 -> Swap `x` and `Some(_)`
3 -> 1 and 2
4 -> Add `let`

### PR DESCRIPTION
**Pull request by chenyukang**
Fixes #103587

### CODE REVIEW

File: compiler/rustc_infer/src/infer/error_reporting/mod.rs [hunk 1]
```diff
@@ -2356,6 +2364,66 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
         diag
     }
 
+    fn suggest_let_for_letchains(
+        &self,
+        err: &mut DiagnosticBuilder<'_, ErrorGuaranteed>,
```
Comment (compiler-errors) [inline compiler/rustc_infer/src/infer/error_reporting/mod.rs:2369]:
```suggestion
        err: &mut Diagnostic,
```

File: compiler/rustc_infer/src/infer/error_reporting/mod.rs [hunk 2]
```diff
@@ -2356,6 +2364,66 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
         diag
     }
 
+    fn suggest_let_for_letchains(
```
Comment (compiler-errors) [inline compiler/rustc_infer/src/infer/error_reporting/mod.rs:2367]: Please comment what this function is doing -- I'm not sure if I understand it.

File: compiler/rustc_infer/src/infer/error_reporting/mod.rs [hunk 3]
```diff
@@ -2356,6 +2364,72 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
...
+                hir::intravisit::walk_body(self, body);
+            }
+        }
+
+        let mut visitor = IfVisitor { err_span: span, found_if: false, result: false };
+        visitor.visit_body(&body);
+        if visitor.result {
+                err.span_suggestion_verbose(
+                    span.shrink_to_lo(),
+                    "consider adding `let`",
```
Comment (davidtwco) [inline compiler/rustc_infer/src/infer/error_reporting/mod.rs:2425]: This could be translatable using a subdiagnostic type.

**Comment by jackh726** [comment]
LGTM with subdiagnostic added

File: compiler/rustc_infer/src/infer/error_reporting/mod.rs (hunk 3 above)
Comment (chenyukang) [reply compiler/rustc_infer/src/infer/error_reporting/mod.rs]: updated!

**Review by davidtwco — APPROVED** [review]

**Comment by Manishearth** [comment]
@bors r-

https://github.com/rust-lang/rust/pull/104176#issuecomment-1308130849

**Comment by chenyukang** [comment]
> @bors r-
> 
> [#104176 (comment)](https://github.com/rust-lang/rust/pull/104176#issuecomment-1308130849)

@Manishearth  How to run this single test case in local?
Seems this command not working:

```bash
x test ./src/tools/clippy/tests/ui/crashes/ice-6250.rs --bless
```

**Comment by Manishearth** [comment]
That ought to work, perhaps ask on Zulip? I wouldn't know without seeing more of the error.

**Comment by davidtwco** [comment]
> @Manishearth How to run this single test case in local? Seems this command not working:
> 
> ```shell
> x test ./src/tools/clippy/tests/ui/crashes/ice-6250.rs --bless
> ```

Wouldn't it be more like `x test ./src/tools/clippy/ --test-args ice-6250 --bless`?

**Comment by chenyukang** [comment]
> > @Manishearth How to run this single test case in local? Seems this command not working:
> > ```shell
> > x test ./src/tools/clippy/tests/ui/crashes/ice-6250.rs --bless
> > ```
> 
> Wouldn't it be more like `x test ./src/tools/clippy/ --test-args ice-6250 --bless`?

No, not working...

```console
cat@LAPTOP-V6U0QKD4:~/code/rust$  x test ./src/tools/clippy/ --test-args ice-6250 --bless
Building rustbuild
    Finished dev [unoptimized] target(s) in 0.04s
Building stage0 std artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
    Finished release [optimized + debuginfo] target(s) in 0.21s
Copying stage0 std from stage0 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu / x86_64-unknown-linux-gnu)
Building stage0 compiler artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
    Finished release [optimized + debuginfo] target(s) in 0.20s
Copying stage0 rustc from stage0 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu / x86_64-unknown-linux-gnu)
Assembling stage1 compiler (x86_64-unknown-linux-gnu)
Building stage1 std artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
    Finished release [optimized + debuginfo] target(s) in 0.14s
Copying stage1 std from stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu / x86_64-unknown-linux-gnu)
Building stage1 compiler artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
    Finished release [optimized + debuginfo] target(s) in 0.19s
Copying stage1 rustc from stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu / x86_64-unknown-linux-gnu)
Building stage1 tool clippy-driver (x86_64-unknown-linux-gnu)
    Finished release [optimized + debuginfo] target(s) in 0.15s
Building rustdoc for stage1 (x86_64-unknown-linux-gnu)
    Finished release [optimized + debuginfo] target(s) in 0.15s
    Finished release [optimized + debuginfo] target(s) in 0.18s
     Running unittests src/driver.rs (build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/clippy_driver-613db7c0b8c82598)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 1 filtered out; finished in 0.00s

     Running tests/check-fmt.rs (build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/check_fmt-7dc715d9b3ea3888)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 1 filtered out; finished in 0.00s

     Running tests/compile-test.rs (build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/compile_test-a1006319e4c0efbe)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 3 filtered out; finished in 0.00s

     Running tests/dogfood.rs (build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/dogfood-4d21110a169a32d8)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 1 filtered out; finished in 0.00s

     Running tests/integration.rs (build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/integration-0664f500aaf92777)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running tests/lint_message_convention.rs (build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/lint_message_convention-c7b02f26e4d0bb14)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 1 filtered out; finished in 0.00s

     Running tests/missing-test-files.rs (build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/missing_test_files-cfe393cf71d33d5d)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 1 filtered out; finished in 0.00s

     Running tests/versioncheck.rs (build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/versioncheck-6c900a9e1e2297ff)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 2 filtered out; finished in 0.00s

     Running tests/workspace.rs (build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/workspace-b27e5604761ba5c8)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 1 filtered out; finished in 0.00s

Build completed successfully in 0:00:02
```

**Comment by chenyukang** [comment]
Got it by this way:

```bash
x.py test --help --verbose | grep clippy
```

Then:
```
./x.py test src/tools/clippy --bless
```

**Comment by jyn514** [comment]
Clippy is a special snowflake and uses `TESTNAME` to tell it how to run individual tests instead of an argument to compiletest. I need to integrate it with x.py at some point. For now you can use `TESTNAME=ice-6250 x test src/tools/clippy --bless`.

I think this would also be fixed by https://github.com/rust-lang/rust/pull/103266.

**Pull request merged**